src/mame/sony/news_68k_iop.cpp

```diff
@@ -0,0 +1,1150 @@
+        void min_w(uint8_t data);
+        void motoron_w(uint8_t data);
+        void powoff_w(uint8_t data);
+        void cpureset_w(uint8_t data);
+        uint8_t rtcreg_r();
+        void rtcreg_w(uint8_t data);
+        void rtcsel_w(uint8_t data);
+        void set_rtc_data_bit(uint8_t bit_number, int data);
+        uint32_t scsi_dma_r(offs_t offset, uint32_t mem_mask);
+        void scsi_dma_w(offs_t offset, uint32_t data, uint32_t mem_mask);
+        void scsi_drq_handler(int status);
+
+        // Front panel signal handlers
+        void handle_rts(int data);
+        void handle_dtr(int data);
+        void update_dcd();
+
+        // Platform hardware used by the main processor
+        void cpu_romdis_w(uint8_t data);
+        void mmuen_w(uint8_t data);
+        uint8_t berr_status_r();
+        void astreset_w(uint8_t data);
+        void astset_w(uint8_t data);
+
+        // CPU timer handlers
+        void interval_timer_tick(uint8_t data);
+        TIMER_CALLBACK_MEMBER(bus_error_off_cpu);
+
+    private:
+        // CPUs (2x 68020, but only the main processor has an FPU)
+        required_device<m68020_device> m_iop;	 // I/O Processor (BSP, brings up system)
+        required_device<m68020fpu_device> m_cpu; // Main Processor
+
+        // Memory devices and main CPU's MMU
+        required_device<news_020_mmu_device> m_mmu;
+        required_device<ram_device> m_ram;
+        required_region_ptr<u32> m_eprom;
+        required_region_ptr<u32> m_idrom;
+
+        // Platform hardware
+        required_device<msm58321_device> m_rtc;
+        required_device<pit8253_device> m_interval_timer;
+        required_device<scc8530_device> m_scc_external;
+        required_device<scc8530_device> m_scc_peripheral;
+        required_device<am7990_device> m_net;
+        required_device<upd765a_device> m_fdc;
+        required_device<ncr5380_device> m_scsi;
+        required_device<news_iop_scsi_helper_device> m_scsi_dma;
+        required_ioport m_dip_switch;
+        required_device_array<rs232_port_device, 2> m_serial;
+        std::unique_ptr<u16[]> m_net_ram;
+
+        // IOP IRQ state
+        uint32_t m_iop_intst = 0;
+        uint32_t m_iop_inten = 0;
+        std::map<int, bool> m_iop_int_state = {
+            {INPUT_LINE_IRQ1, false},
+            {INPUT_LINE_IRQ2, false},
+            {INPUT_LINE_IRQ3, false},
+            {INPUT_LINE_IRQ4, false},
+            {INPUT_LINE_IRQ5, false},
+            {INPUT_LINE_IRQ6, false},
+            {INPUT_LINE_IRQ7, false}
+        };
+
+        // CPU IRQ state
+        uint32_t m_cpu_intst = 0;
+        uint32_t m_cpu_inten = 0;
+        std::map<int, bool> m_cpu_int_state = {
+            {INPUT_LINE_IRQ1, false},
+            {INPUT_LINE_IRQ2, false},
+            {INPUT_LINE_IRQ3, false},
+            {INPUT_LINE_IRQ4, false},
+            {INPUT_LINE_IRQ5, false},
+            {INPUT_LINE_IRQ6, false},
+            {INPUT_LINE_IRQ7, false}
+        };
```

Comment:
Do you really need `std::map` for these?  It’s going to make save state support a pain.

Reply:
No, I wrote this in a way that made sense to me when I was untangling the messy code I originally had from experimenting with the IRQ setup for the IOP. But I didn't think about save state support, so I will go back to the drawing board on this one.

Reply:
Eliminated this part entirely, I don't know what I was thinking. Guess I missed that `input_state` existed :D